core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -257,6 +264,37 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  // Changes are initially deferred when using a Raft-based metadata quorum, and they may flip-flop to not
+  // being deferred and being deferred again thereafter as the broker (re)acquires/loses its lease.
+  // Changes are never deferred when using ZooKeeper.  When true, this indicates that we should transition
+  // online partitions to the deferred state if we see a metadata update for that partition.
+  @volatile private var deferringMetadataChanges: Boolean = !config.requiresZookeeper
+  stateChangeLogger.info(s"Metadata changes deferred=$deferringMetadataChanges")
+
+  private def confirmNotDeferringMetadataUpdatesWithZooKeeper(): Unit = {
+    if (deferringMetadataChanges) {
+      throw new IllegalStateException("Partition metadata changes should never be deferred when using ZooKeeper")
+    }
+  }
+
+  def beginMetadataChangeDeferral(): Unit = {
+    replicaStateChangeLock synchronized {
+      if (config.requiresZookeeper) {
+        throw new IllegalStateException("Partition metadata changes can never be deferred when using ZooKeeper")
+      }
+      deferringMetadataChanges = true
+      stateChangeLogger.info(s"Metadata changes are now being deferred")
+    }
+  }
+
+  def endMetadataChangeDeferral(): Unit = {
+    replicaStateChangeLock synchronized {
+      // TODO: implement for Raft-based metadata log messages
```

Comment:
nit: we can leave the TODO out. 

Also, it might be worth checking `config.requiresZookeeper` as we do in `beginMetadataChangeDeferral`. If nothing else, it serves as explicit documentation.